src/text-fragment-utils.js

```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
     range.setEnd(endNode, endOffset);
     try {
       range.surroundContents(mark);
+      return [mark];
     } catch {
-      // Text to highlight does not contain entire DOM nodes.
-      // Need to extend the highlighted selection to entire nodes (Ex: entire links).
+      // Text to highlight does not contain entire DOM elements.
+      // Need to create multiple <mark> elements to highlight the entire selection.
       const commonAncestor = range.commonAncestorContainer;
-      while (startNode.parentNode !== commonAncestor) {
-        startNode = startNode.parentNode;
-        range.setStartBefore(startNode);
-      }
-      while (endNode.parentNode !== commonAncestor) {
-        endNode = endNode.parentNode;
-        range.setEndAfter(endNode);
+      const startMarks = [];
+      let firstNode = startNode;
+      startMarks.push(
+        createMarkFor(firstNode, {
+          start: { node: firstNode, offset: startOffset },
+        }),
+      );
+      firstNode = startMarks[0];
+      while (firstNode.parentNode !== commonAncestor) {
```

Comment:
Could you add a comment for this while loop explaining what the goal of walking the tree here is?